client/asset/eth/eth.go

```diff
@@ -0,0 +1,534 @@
+func (d *Driver) Setup(cfg *asset.WalletConfig, logger dex.Logger, network dex.Network) (asset.Wallet, error) {
+	return NewWallet(cfg, logger, network)
+}
+
+// DecodeCoinID creates a human-readable representation of a coin ID for Ethereum.
+func (d *Driver) DecodeCoinID(coinID []byte) (string, error) {
+	return dexeth.CoinIDToString(coinID)
+}
+
+// Info returns basic information about the wallet and asset.
+func (d *Driver) Info() *asset.WalletInfo {
+	return WalletInfo
+}
+
+// rawWallet is an unexported return type from the eth client. Watch for changes at
+// https://github.com/ethereum/go-ethereum/blob/c503f98f6d5e80e079c1d8a3601d188af2a899da/internal/ethapi/api.go#L227-L253
+type rawWallet struct {
+	URL      string             `json:"url"`
+	Status   string             `json:"status"`
+	Failure  string             `json:"failure,omitempty"`
+	Accounts []accounts.Account `json:"accounts,omitempty"`
+}
+
+// ethFetcher represents a blockchain information fetcher. In practice, it is
+// satisfied by rpcclient. For testing, it can be satisfied by a stub.
+type ethFetcher interface {
+	accounts() []*accounts.Account
+	addPeer(ctx context.Context, peer string) error
+	balance(ctx context.Context, acct *accounts.Account) (*big.Int, error)
+	bestBlockHash(ctx context.Context) (common.Hash, error)
+	block(ctx context.Context, hash common.Hash) (*types.Block, error)
+	blockNumber(ctx context.Context) (uint64, error)
+	connect(ctx context.Context, node *node.Node, contractAddr common.Address) error
+	importAccount(pw string, privKeyB []byte) (*accounts.Account, error)
+	listWallets(ctx context.Context) ([]rawWallet, error)
+	lock(ctx context.Context, acct *accounts.Account) error
+	nodeInfo(ctx context.Context) (*p2p.NodeInfo, error)
+	pendingTransactions(ctx context.Context) ([]*types.Transaction, error)
+	peers(ctx context.Context) ([]*p2p.PeerInfo, error)
+	transactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error)
+	sendTransaction(ctx context.Context, tx map[string]string) (common.Hash, error)
+	shutdown()
+	syncProgress(ctx context.Context) (*ethereum.SyncProgress, error)
+	unlock(ctx context.Context, pw string, acct *accounts.Account) error
+}
+
+// Check that ExchangeWallet satisfies the asset.Wallet interface.
+var _ asset.Wallet = (*ExchangeWallet)(nil)
+
+// ExchangeWallet is a wallet backend for Ethereum. The backend is how the DEX
+// client app communicates with the Ethereum blockchain and wallet. ExchangeWallet
+// satisfies the dex.Wallet interface.
+type ExchangeWallet struct {
+	// 64-bit atomic variables first. See
+	// https://golang.org/pkg/sync/atomic/#pkg-note-BUG
+	tipAtConnect int64
+
+	ctx       context.Context // the asset subsystem starts with Connect(ctx)
+	node      ethFetcher
+	log       dex.Logger
+	tipChange func(error)
+
+	internalNode *node.Node
+
+	tipMtx     sync.RWMutex
+	currentTip *types.Block
+
+	acctMtx sync.RWMutex
+	acct    *accounts.Account
```

Comment:
Yes, originally I had thought this would be imported or something, but can be created deterministically with our client seed, after that initial work is done.

Reply:
Removing the mutex anyhow.